### PR TITLE
keep holding layer map lock inside `compact_level0_phase1`

### PR DESCRIPTION
**Pull request by problame**
This is preliminary work for/from #4220 (async `Layer::get_value_reconstruct_data`).

That plan involves turning Timeline::layers into a `tokio::sync::RwLock`.

Before this patch, `compact_level0_phase1` would drop and re-acquire `Timeline::layers` (`read().unwrap()`).

We can't have that if we switch to an async RwLock because tokio::sync::RwLock's read guard is `!Send`.
Which makes the compaction task future `!Send`.
Which doesn't compile because the compaction task runs inside a `task_mgr` task, and `task_mgr` requires the task futures to be `Send`.

I think the performance impact of holding the lock for longer will be minimal as the number of L0s should be bounded, and so, any of the prep work that we're doing in the section where we previously didn't hold the lock should be quite fast.

### CODE REVIEW

**Review by hlinnaka — APPROVED** [review]
> The far worse part of this function in terms of lock contention / performance is the fact that we hold the read lock while we write out the L1s.

Huh, no. We have this later in the function:

```
        drop(layers);
        let mut holes = heap.into_vec();
        holes.sort_unstable_by_key(|hole| hole.key_range.start);
        let mut next_hole = 0; // index of next hole in holes vector

        // Merge the contents of all the input delta layers into a new set
        // of delta layers, based on the current partitioning.
 ```
After that, we write out the L1 layer files.

This makes me a bit uneasy, we should be reducing lock contention, not increasing it. Maybe some tests would make us more confident? The lock contention is probably unnoticeable after all, so I'm clicking "approve", but please consider if there's some easy testing or something we could do to support that.

**Comment by skyzh** [comment]
I believe the guard should be send because tokio allows you to await inside the critical section. This might be caused by some other things. I remembered I got this issue and fixed it by wrapping the lock region with {}.

File: pageserver/src/tenant/timeline.rs
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3326]: Note that we're in a `spawn_blocking` sync function here. So, we could just use `blocking_read`.

But, that wastes a spawn-blocking-thread-pool slot.

The patch that converts `Timeline::layers` to an async rwlock will move the locked `layers` into the `spawn_blocking`.

**Comment by problame** [comment]
So, the compilation error inside #4441 if I leave out this patch is

```
     = help: the trait `std::marker::Send` is not implemented for `dyn std::iter::Iterator<Item = (repository::Key, utils::lsn::Lsn, u64)>`
note: future is not `Send` as this value is used across an await
    --> pageserver/src/tenant/timeline.rs:3485:40
     |
3467 |         let mut all_keys_iter = itertools::process_results(
     |             ----------------- has type `itertools::KMergeBy<std::boxed::Box<dyn std::iter::Iterator<Item = (repository::Key, utils::lsn::Lsn, u64)>>, [closure@pageserver/src/tenant/timeline.rs:3470:37: 3470:43]>` which is not `Send`
```

because we hold that across the second `read().await`.

This can be fixed by `Arc`wrapping the `deltas_to_compact` so that the `all_keys_iter` and the `all_values_iter` own it.

Gut feeling says that's more expensive than the longer hold duration for the `read` lock.

**Comment by skyzh** [comment]
caused by the type we defined:

```
pub type LayerKeyIter<'i> = Box<dyn Iterator<Item = (Key, Lsn, u64)> + 'i>;
```

should be 

```
pub type LayerKeyIter<'i> = Box<dyn Iterator<Item = (Key, Lsn, u64)> + 'i + Send>;
```

instead. I think it's okay to get this merged, as after the immutable layer refactor we don't need to take locks anyways :)

**Comment by problame** [comment]
Good catch, fixing that yields

```
    Checking pageserver v0.1.0 (/Users/cs/src/neon/pageserver)
error[E0277]: `std::sync::RwLockReadGuard<'_, tenant::storage_layer::delta_layer::DeltaLayerInner>` cannot be sent between threads safely
   --> pageserver/src/tenant/storage_layer/delta_layer.rs:422:25
    |
422 |             Ok(iter) => Box::new(iter),
    |                         ^^^^^^^^^^^^^^ `std::sync::RwLockReadGuard<'_, tenant::storage_layer::delta_layer::DeltaLayerInner>` cannot be sent between threads safely
    |
    = help: within `tenant::storage_layer::delta_layer::DeltaValueIter<'_>`, the trait `std::marker::Send` is not implemented for `std::sync::RwLockReadGuard<'_, tenant::storage_layer::delta_layer::DeltaLayerInner>`
note: required because it appears within the type `Adapter<'_>`
   --> pageserver/src/tenant/storage_layer/delta_layer.rs:943:8
    |
943 | struct Adapter<'a>(RwLockReadGuard<'a, DeltaLayerInner>);
    |        ^^^^^^^
note: required because it appears within the type `BlockCursor<Adapter<'_>>`
   --> pageserver/src/tenant/block_io.rs:68:12
    |
68  | pub struct BlockCursor<R>
    |            ^^^^^^^^^^^
note: required because it appears within the type `DeltaValueIter<'_>`
   --> pageserver/src/tenant/storage_layer/delta_layer.rs:937:8
    |
937 | struct DeltaValueIter<'a> {
    |        ^^^^^^^^^^^^^^
    = note: required for the cast from `tenant::storage_layer::delta_layer::DeltaValueIter<'_>` to the object type `dyn Iterator<Item = Result<(Key, Lsn, Value), Error>> + Send`
    = note: the full name for the object type has been written to '/Users/cs/src/neon/target/debug/deps/pageserver-4dd2d7d39f9b303e.long-type-153329869280982885.txt'
```


The offending code is

```
  let inner = self
            .load(LayerAccessKind::KeyIter, ctx)
            .context("load delta layer")?;
        Ok(match DeltaValueIter::new(inner) {
```

**Comment by problame** [comment]
Wondering whether the `inner` can just be a `Once` ?

**Comment by skyzh** [comment]
An intuitive fix might be changing DeltaLayerIter to hold `&'a DeltaLayerInner` instead of `RwLockGuard`. RwLockGuard cannot be Send as Linux requires lock / unlock happens in the same thread.

**Comment by skyzh** [comment]
Well in this case, probably we still need this PR 🤣

**Comment by problame** [comment]
Yeah, I'm converting the `inner` to `OnceCell`, looks promising, will open a new PR for it though

**Pull request closed without merging**
